File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
def permutations(arr, m):
    result = []
    permu = []

    visited = [False] * len(arr)

    def backtrack():
        if len(permu) == m:
            result.append(permu[:])

        for i in range(len(arr)):
            if not visited[i]:
                visited[i] = True

                permu.append(arr[i])
                backtrack()
                permu.pop()

                visited[i] = False

    backtrack()

    return result

n, m = map(int, input().split())

arr = list(range(1, n + 1))

result = permutations(arr, m)

for res in result:
    print(*res)

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.cc
#include <bits/stdc++.h>
using namespace std;

int n, m;

vector<vector<int>> permutation(const vector<int>& arr, int k) {
	vector<int> temp;
	vector<vector<int>> res;
	vector<bool> visited(arr.size(), false);

	function<void()> backtrack = [&]() {
		if (temp.size() == k) {
			res.push_back(temp);
			return;
		}

		for (int i = 0; i < arr.size(); ++i) {
			if (visited[i])
				continue;
			
			visited[i] = true;
			temp.push_back(arr[i]);
			backtrack();
			temp.pop_back();
			visited[i] = false;
		}
	};

	backtrack();

	return res;
}

int main() {
	cin >> n >> m;

	vector<int> arr;

	for (int num = 1; num <= n; ++num) {
		arr.push_back(num);
	}

	vector<vector<int>> result = permutation(arr, m);

	for (auto vec : result) {
		for (auto v : vec) {
			cout << v << " ";
		}
		cout << "\n";
	}

	return 0;
}